File: app/models/itinerary.py
import uuid
from datetime import datetime
from sqlalchemy import (
    Column,
    String,
    Text,
    Integer,
    Boolean,
    Numeric,
    DateTime,
    Date,
    Time,
    ForeignKey,
    ARRAY,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.core.database import Base


class TripRequest(Base):
    """User trip request input."""

    __tablename__ = "trip_requests"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True))

    # Basic Info
    destination_city = Column(String(100))
    start_date = Column(Date)
    end_date = Column(Date)

    # Persona Selection
    group_type = Column(String(50))
    group_size = Column(Integer)
    has_kids = Column(Boolean, default=False)
    kids_ages = Column(ARRAY(Integer))
    has_seniors = Column(Boolean, default=False)

    # Preferences
    vibes = Column(ARRAY(String))  # selected vibes
    budget_level = Column(Integer)  # 1-5
    daily_budget = Column(Numeric(10, 2))
    pacing = Column(String(20))  # slow, moderate, fast

    # Constraints
    mobility_constraints = Column(ARRAY(String))
    dietary_restrictions = Column(ARRAY(String))

    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    itineraries = relationship("Itinerary", back_populates="trip_request", cascade="all, delete-orphan")


class Itinerary(Base):
    """Generated itinerary."""

    __tablename__ = "itineraries"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    trip_request_id = Column(UUID(as_uuid=True), ForeignKey("trip_requests.id", ondelete="CASCADE"))

    # Metadata
    total_estimated_cost = Column(Numeric(10, 2))
    generation_method = Column(String(50))  # rag_v1, rag_v2, etc.

    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    trip_request = relationship("TripRequest", back_populates="itineraries")
    days = relationship("ItineraryDay", back_populates="itinerary", cascade="all, delete-orphan")


class ItineraryDay(Base):
    """A single day in an itinerary."""

    __tablename__ = "itinerary_days"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    itinerary_id = Column(UUID(as_uuid=True), ForeignKey("itineraries.id", ondelete="CASCADE"))
    day_number = Column(Integer)
    date = Column(Date)
    theme = Column(String(100))  # "Beach & Relaxation", "Cultural Exploration"

    # Daily summary
    estimated_cost = Column(Numeric(10, 2))
    pacing_score = Column(Numeric(3, 2))  # actual pacing achieved

    # Relationships
    itinerary = relationship("Itinerary", back_populates="days")
    items = relationship("ItineraryItem", back_populates="day", cascade="all, delete-orphan")


class ItineraryItem(Base):
    """A single item (activity/meal) in an itinerary day."""

    __tablename__ = "itinerary_items"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    itinerary_day_id = Column(UUID(as_uuid=True), ForeignKey("itinerary_days.id", ondelete="CASCADE"))
    poi_id = Column(UUID(as_uuid=True), ForeignKey("pois.id", ondelete="SET NULL"))

    sequence_order = Column(Integer)
    start_time = Column(Time)
    end_time = Column(Time)

    # Why this was selected (explainability)
    selection_reason = Column(Text)
    persona_match_score = Column(Numeric(3, 2))

    # Logistics
    travel_time_from_previous = Column(Integer)  # minutes
    travel_mode = Column(String(20))  # walk, drive, transit

    # Relationships
    day = relationship("ItineraryDay", back_populates="items")
    poi = relationship("POI")

File: flutter_app/lib/models/itinerary.dart
import 'poi.dart';

class TripRequest {
  final String? userId;
  final String destinationCity;
  final DateTime startDate;
  final DateTime endDate;
  final String groupType;
  final int groupSize;
  final bool hasKids;
  final bool hasSeniors;
  final List<String> vibes;
  final int budgetLevel;
  final double? dailyBudget;
  final String pacing;

  TripRequest({
    this.userId,
    required this.destinationCity,
    required this.startDate,
    required this.endDate,
    required this.groupType,
    this.groupSize = 2,
    this.hasKids = false,
    this.hasSeniors = false,
    required this.vibes,
    this.budgetLevel = 3,
    this.dailyBudget,
    this.pacing = 'moderate',
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'destination_city': destinationCity,
      'start_date': startDate.toIso8601String().split('T')[0],
      'end_date': endDate.toIso8601String().split('T')[0],
      'group_type': groupType,
      'group_size': groupSize,
      'has_kids': hasKids,
      'has_seniors': hasSeniors,
      'vibes': vibes,
      'budget_level': budgetLevel,
      'daily_budget': dailyBudget,
      'pacing': pacing,
    };
  }
}

class Itinerary {
  final String id;
  final String? tripRequestId;
  final double? totalEstimatedCost;
  final String? generationMethod;
  final DateTime? createdAt;
  final List<ItineraryDay> days;
  final String? cityName;

  Itinerary({
    required this.id,
    this.tripRequestId,
    this.totalEstimatedCost,
    this.generationMethod,
    this.createdAt,
    required this.days,
    this.cityName,
  });

  factory Itinerary.fromJson(Map<String, dynamic> json) {
    return Itinerary(
      id: json['id'] ?? '',
      tripRequestId: json['trip_request_id'],
      totalEstimatedCost: json['total_estimated_cost']?.toDouble(),
      generationMethod: json['generation_method'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      days: (json['days'] as List<dynamic>?)
              ?.map((d) => ItineraryDay.fromJson(d))
              .toList() ??
          [],
      cityName: json['city_name'],
    );
  }

  int get totalDays => days.length;

  String get costText {
    if (totalEstimatedCost == null) return 'N/A';
    return '\$${totalEstimatedCost!.toStringAsFixed(0)}';
  }
}

class ItineraryDay {
  final String id;
  final int dayNumber;
  final DateTime? date;
  final String? theme;
  final double? estimatedCost;
  final double? pacingScore;
  final List<ItineraryItem> items;

  ItineraryDay({
    required this.id,
    required this.dayNumber,
    this.date,
    this.theme,
    this.estimatedCost,
    this.pacingScore,
    required this.items,
  });

  factory ItineraryDay.fromJson(Map<String, dynamic> json) {
    return ItineraryDay(
      id: json['id'] ?? '',
      dayNumber: json['day_number'] ?? 1,
      date: json['date'] != null ? DateTime.parse(json['date']) : null,
      theme: json['theme'],
      estimatedCost: json['estimated_cost']?.toDouble(),
      pacingScore: json['pacing_score']?.toDouble(),
      items: (json['items'] as List<dynamic>?)
              ?.map((i) => ItineraryItem.fromJson(i))
              .toList() ??
          [],
    );
  }

  String get dateText {
    if (date == null) return 'Day $dayNumber';
    final months = [
      'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
      'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
    ];
    return '${months[date!.month - 1]} ${date!.day}';
  }
}

class ItineraryItem {
  final String id;
  final String? poiId;
  final int sequenceOrder;
  final String? startTime;
  final String? endTime;
  final String? selectionReason;
  final double? personaMatchScore;
  final int? travelTimeFromPrevious;
  final String? travelMode;
  final POI? poi;

  ItineraryItem({
    required this.id,
    this.poiId,
    required this.sequenceOrder,
    this.startTime,
    this.endTime,
    this.selectionReason,
    this.personaMatchScore,
    this.travelTimeFromPrevious,
    this.travelMode,
    this.poi,
  });

  factory ItineraryItem.fromJson(Map<String, dynamic> json) {
    return ItineraryItem(
      id: json['id'] ?? '',
      poiId: json['poi_id'],
      sequenceOrder: json['sequence_order'] ?? 0,
      startTime: json['start_time'],
      endTime: json['end_time'],
      selectionReason: json['selection_reason'],
      personaMatchScore: json['persona_match_score']?.toDouble(),
      travelTimeFromPrevious: json['travel_time_from_previous'],
      travelMode: json['travel_mode'],
      poi: json['poi'] != null ? POI.fromJson(json['poi']) : null,
    );
  }

  String get timeRange {
    if (startTime == null || endTime == null) return '';
    return '$startTime - $endTime';
  }

  String get travelText {
    if (travelTimeFromPrevious == null || travelTimeFromPrevious == 0) {
      return '';
    }
    final mode = travelMode ?? 'walk';
    final icon = mode == 'transit' ? 'transit' : 'walk';
    return '${travelTimeFromPrevious}min $icon';
  }
}
